p2p/stream/common/ratelimiter/ratelimiter.go

```diff
@@ -0,0 +1,60 @@
+package ratelimiter
+
+import (
+	"sync"
+
+	sttypes "github.com/harmony-one/harmony/p2p/stream/types"
+	"go.uber.org/ratelimit"
+)
+
+// RateLimiter is the interface to limit the incoming request.
+// The purpose of rate limiter is to prevent the node from running out of resource
+// for consensus on DDoS attacks.
+type RateLimiter interface {
+	LimitRequest(stid sttypes.StreamID)
+}
+
+// rateLimiter is the implementation of RateLimiter.
+// The rateLimiter limit request rate:
+//   1. For global stream requests
+//   2. For requests from a stream
+// TODO: make request weighted in rate limiter
+type rateLimiter struct {
+	globalLimiter ratelimit.Limiter
+
+	streamRate int
+	limiters   map[sttypes.StreamID]ratelimit.Limiter
+
+	lock sync.RWMutex
+}
+
+// NewRateLimiter creates a new rate limiter
+func NewRateLimiter(global, single int) RateLimiter {
+	return &rateLimiter{
+		globalLimiter: ratelimit.New(global),
+
+		streamRate: single,
+		limiters:   make(map[sttypes.StreamID]ratelimit.Limiter),
+	}
+}
+
+func (rl *rateLimiter) LimitRequest(stid sttypes.StreamID) {
+	rl.lock.RLock()
+	limiter, ok := rl.limiters[stid]
```

Comment:
limiter is per stream id, will you release the limiter from the map?

Reply:
Good point. Let me do it.

Reply:
After looking into the prometheus for a while, I think I will add prometheus support after the merge of PRs for the following reason:
1. The prometheus metrics will not effect the key feature. It's a better-to-have feature.
2. Prometheus metrics can be added in a separate PR which will make the changes more clean.
So I created [an issue](https://github.com/harmony-one/harmony/issues/3555) to be executed after the review of this entire merge.

Reply:
Resolved. Please review again.